spec/index.bs

```diff
@@ -2564,6 +2569,97 @@ If a format is listed in [[#plain-color-formats]] with {{GPUTextureUsage/RENDER_
+
+Note: External textures *can* be implemented without creating a copy of the imported canvas/video,
+but a user agent may not always implement this for all cases.
+
+<script type=idl>
+[Exposed=Window]
+interface GPUExternalTexture {
+};
+GPUExternalTexture includes GPUObjectBase;
+</script>
+
+{{GPUExternalTexture}} has the following internal slots:
+
+<dl dfn-type=attribute dfn-for="GPUExternalTexture">
+    : <dfn>\[[destroyed]]</dfn>, of type `boolean`
+    ::
+        Indicates whether the object has been destroyed (can no longer be used).
+        Initially set to `false`.
+
+        <!-- Move this to an internal object if GPUExternalTexture becomes Serializable. -->
+</dl>
+
+### Importing External Textures ### {#external-texture-creation}
+
+An external texture is created from an external video object
+using {{GPUDevice/importExternalTexture()}}.
+
+External textures are destroyed automatically, as a microtask,
+instead of manually or upon garbage collection like other resources.
+
+<script type=idl>
+dictionary GPUExternalTextureDescriptor : GPUObjectDescriptorBase {
+    required HTMLVideoElement source;
+};
+</script>
+
+<dl dfn-type=method dfn-for=GPUDevice>
+    : <dfn>importExternalTexture(descriptor)</dfn>
+    ::
+        Creates a {{GPUExternalTexture}} wrapping the provided image source.
+
+        <div algorithm=GPUDevice.importExternalTexture>
+            **Called on:** {{GPUDevice}} this.
+
+            **Arguments:**
+            <pre class=argumentdef for="GPUDevice/importExternalTexture(descriptor)">
+                |descriptor|: Provides the external image source object (and any creation options).
+            </pre>
+
+            **Returns:** {{GPUExternalTexture}}
+
+            1. Let |source| be |descriptor|.{{GPUExternalTextureDescriptor/source}}.
+
+            1. If |source| <l spec=html>[=is not origin-clean=]</l>,
+                throw a {{SecurityError}} and stop.
+
+            1. Let |result| be a new {{GPUExternalTexture}} object wrapping a copy of the current
+                image contents of |source|.
+
+                Note: This is described as a copy, but may be implemented without an actual copy.
+
+            1. [=Queue a microtask=] to set |result|.{{GPUExternalTexture/[[destroyed]]}} to `true`,
+                releasing the underlying resource.
+
+                Issue: Is this too restrictive?
+        </div>
+</dl>
+
+### Sampling External Textures ### {#external-texture-sampling}
+
+External textures are represented in WGSL with `texture_external` and may be sampled using
+usual texture load/sample functions, plus `textureLoadLuminance` and `textureSampleLuminance`.
+
+The provided `sampler` is used to sample any underlying texture(s).
+If there is one RGBA underlying texture, this behaves as usual.
+If there are several underlying textures (Y+U+V or Y+UV), the sampler is used to sample each
+underlying texture separately, prior to conversion from YUV to RGBA or luminance.
```

Comment:
We previously discussed leaving things open to allow implementations to still filter after conversion (if they wanted to), but the API surface in this PR doesn't allow it: the sampling parameters are baked into the sampler object, so they can't be passed in through the external texture's metadata UBO. Hence I've written the result down here concretely.

But if we DO want post-conversion filtering to be possible with 0-copy YUV sources, we should change the API to allow this somehow.

Reply:
I think the post-conversion filtering can still be done with sampler with the same amount of work as when you don't have a sampler:

 - Take the fractional and integer parts of the texture coordinates (after multiplication by the texture size)
 - Sample the square starting at integer part with the sampler
 - Mix using the fractional part.

Reply:
But how do you know whether to use linear or nearest filtering? I don't think the shader can introspect that info from a sampler object.

Reply:
Ah yeah that makes sense. If we wanted to allow that we would need to add the sampler to extrenal_texture creation.